internal/lsp/server.go

```diff
@@ -800,6 +803,56 @@ func (l *LanguageServer) handleTextDocumentDidChange(
 	return struct{}{}, nil
 }
 
+func (l *LanguageServer) handleTextDocumentDidSave(
+	ctx context.Context,
+	_ *jsonrpc2.Conn,
+	req *jsonrpc2.Request,
+) (result any, err error) {
+	var params types.TextDocumentDidSaveParams
+	if err := json.Unmarshal(*req.Params, &params); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal params: %w", err)
+	}
+
+	if params.Text != nil && l.loadedConfig != nil {
+		if !strings.Contains(*params.Text, "\r\n") {
+			return struct{}{}, nil
+		}
+
+		enabled, err := linter.NewLinter().WithUserConfig(*l.loadedConfig).DetermineEnabledRules(ctx)
+		if err != nil {
+			l.logError(fmt.Errorf("failed to determine enabled rules: %w", err))
+
+			return struct{}{}, nil
+		}
+
+		formattingEnabled := false
+
+		for _, rule := range enabled {
+			if rule == "opa-fmt" || rule == "use-rego-v1" {
+				formattingEnabled = true
+
+				break
+			}
+		}
+
+		if formattingEnabled {
+			resp := types.ShowMessageParams{
+				Type:    2, // warning
+				Message: "CRLF line ending detected. Please change editor setting to use LF for line endings.",
+			}
+
+			err := l.conn.Notify(ctx, "window/showMessage", resp)
```

Comment:
I remember there was an issue in `handleWorkspaceExecuteCommand` where we had blocking issues sending notify messages mid request response. I guess this isn't an issue here if it works, but something to double check.

We don't have any other Notify calls that are sent at request time. However, I think this is fine here as the response isn't needed/blocking at the client for this operation - it would seem.

Reply:
Yeah, I just interpreted "notification" to mean "notify" in the spec sense: https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#window_showMessage

But it probably just means "notification" as in the UI.. do you think I should use a `Call` here instead?

Reply:
But yeah, this does seem to work reliably.